tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py

```diff
@@ -254,29 +254,45 @@ def __init__(self, key, env_var, value, defaulted_message):
         f.write(config)
 
 
-def wait_for_model(rest_port, model_name, timeout_seconds, wait_interval_seconds=5):
+def wait_for_model(rest_port, model_name, timeout_seconds, pid):
     tfs_url = "http://localhost:{}/v1/models/{}".format(rest_port, model_name)
 
-    with timeout(timeout_seconds):
-        while True:
-            try:
-                session = requests.Session()
-                retries = Retry(total=9, backoff_factor=0.1)
-                session.mount("http://", requests.adapters.HTTPAdapter(max_retries=retries))
-                log.info("Trying to connect with model server: {}".format(tfs_url))
-                response = session.get(tfs_url)
-                log.info(response)
-                if response.status_code == 200:
-                    versions = json.loads(response.content)["model_version_status"]
-                    if all(version["state"] == "AVAILABLE" for version in versions):
-                        break
-            except (
-                ConnectionRefusedError,
-                NewConnectionError,
-                MaxRetryError,
-                requests.exceptions.ConnectionError,
-            ):
-                log.warning("model: {} is not available yet ".format(tfs_url))
-                time.sleep(wait_interval_seconds)
-
-    log.info("model: {} is available now".format(tfs_url))
+    try:
+        session = requests.Session()
+        backoff_factor = 0.1
+        # sleep = {backoff factor} * (2 ^ ({number of retries left} - 1))
+        retry_count = retry_from_timeout(timeout_seconds, backoff_factor)
+        retries = Retry(total=retry_count, backoff_factor=backoff_factor)
+        session.mount("http://", requests.adapters.HTTPAdapter(max_retries=retries))
+        log.info(
+            "Trying to connect with model server: {} with timeout : {} and retry : {}".format(tfs_url, timeout_seconds,
+                                                                                              retry_count))
+        response = session.get(tfs_url, timeout=0.1)
+        log.info(response)
+        if response.status_code == 200:
+            versions = json.loads(response.content)["model_version_status"]
+        if all(version["state"] == "AVAILABLE" for version in versions):
+            return
+        raise MultiModelException(408, "Timed out after {} seconds".format(timeout_seconds), pid)
+    except (
+            ConnectionRefusedError,
+            NewConnectionError,
+            MaxRetryError,
+            requests.exceptions.ConnectionError,
+    ):
+        raise MultiModelException(408, "Timed out after {} seconds".format(timeout_seconds), pid)
+
+
+def retry_from_timeout(timeout_seconds, backoff_factor):
```

Comment:
with timeout(timeout_seconds):
        while True:
            try:
                session = requests.Session()
                retries = Retry(total=9, backoff_factor=0.1)
                session.mount("http://", requests.adapters.HTTPAdapter(max_retries=retries))
                response = session.get(tfs_url)
                ...
            except (
                ...

The above is the original code. During the `retry`, if `with timeout` throw exception, it will be caught by `retry` and increase retry counter by 1. In other words, if tfs fails to start, this function will retry indefinitely. 

Since requests.Session() does not provide a total timeout. the walk around is base on the total timeout to calculate retry count and limit each request with timeout 100ms.

Reply:
Yes, the `while True` loop is broken, I agree. I get what you are trying to calculate here. I'm saying that the calculation isn't right anyway, because it only counts the time between requests, not including the actual request time itself. So your calculated retry count is going to be too high if the model load requests take a significant amount of time, making it kind of pointless.

Reply:
Sync up offline. The following is the problem statement:

In order to use `multiprocess.manager.dict` to we have to use gthread instead of gevent. 
Here is the issue using gevent with `multiprocess.manager.dict` : https://github.com/benoitc/gunicorn/issues/1407

However, timeout using signal not work with `gthread` with following error message:
```
2023-04-04 16:05:22 [FALCON] [ERROR] GET /increment/abc => Traceback (most recent call last):
  File "falcon/app.py", line 365, in falcon.app.App.__call__
  File "/sagemaker_tmp/test.py", line 53, in on_get
    with timeout(2):
  File "/usr/local/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/sagemaker_tmp/test.py", line 43, in timeout
    signal.signal(signal.SIGALRM, _raise_timeout_error)
  File "/usr/local/lib/python3.9/signal.py", line 56, in signal
    handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
ValueError: signal only works in main thread of the main interpreter
```

For now, we are calculating timeout. Added comment about small delta we might have.